File: 백준/Silver/16948. 데스 나이트/데스 나이트.py
import sys
from collections import deque

input = sys.stdin.readline

n = int(input())
arr = [int(i) for i in input().split()]

r1, c1, r2, c2 = arr

visited = set()

d = [(-2, -1), (-2, 1), (0, -2), (0, 2), (2, -1), (2, 1)]


def bfs():
    que = deque()
    que.append((r1, c1, 0))
    visited.add((r1, c1))

    while que:
        r, c, cnt = que.popleft()

        if r == r2 and c == c2:
            return cnt

        for dr, dc in d:
            nr = r + dr
            nc = c + dc

            if nr >= n or nr < 0 or nc >= n or nc < 0 or (nr, nc) in visited:
                continue

            visited.add((nr, nc))
            que.append((nr, nc, cnt + 1))

    return -1


def main():
    return bfs()


print(main())

File: 백준/Silver/16948. 데스 나이트/데스 나이트.js
const fs = require("fs");
const read = fs.readFileSync("dev/stdin").toString().trim();

// const read = `6
// 5 1 0 5`;

var [n, arr] = read.split("\n");

n = Number(n);
let [r1, c1, r2, c2] = arr.split(" ").map(Number);
let d = [
  [-2, -1],
  [-2, 1],
  [0, -2],
  [0, 2],
  [2, -1],
  [2, 1],
];
visited = new Set();

const bfs = () => {
  que = [];
  que.push([r1, c1, 0]);

  visited.add(r1 * n + c1);

  while (que.length) {
    let [r, c, cnt] = que.shift();
    if (r == r2 && c == c2) return cnt;

    for (let dd of d) {
      let nr = r + dd[0];
      let nc = c + dd[1];

      if (nr < 0 || nr >= n || nc < 0 || nc >= n || visited.has(nr * n + nc))
        continue;

      visited.add(nr * n + nc);
      que.push([nr, nc, cnt + 1]);
    }
  }

  return -1;
};

const main = () => {
  return bfs();
};

console.log(main());
